text/0000-danger-zone.md

```diff
@@ -0,0 +1,64 @@
+# Meta
+[meta]: #meta
+- Name: Read/Write Platform Volume Mount
+- Start Date: 2020-06-02
+- Author(s): nebhale
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+Buildpacks should have access to a read/write file system location provided by the platform.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, every part of the filesystem available during `detect` and every part of the filesystem other than layers during `build` are read-only.  This satisifies most use-cases, but specifically does not support the possibility of multiple builds sharing a communal filesystem location.  While this would likely to be a disaster in many situations, especially in production environments, being able to share a filesystem between a development laptop and a build is _very_ useful.  For example, the single longest bit of building a Java application from source is the initial population of a Maven or Gradle cache.  Hundreds and perhaps thousands of artifacts are downlaoded to do the very first build and subsequently never need to be downloaded again.  Caching these artifacts in a `cache = true` layer helps, but doesn't solve the first build speed problem and results in a pretty poor demo experience.  If a user could mount their `~/.m2` or `~/.gradle` folders into a safe location within the build container, accepting the overall risk of such a choice, the experience would be vastly superior.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This provides a high level overview of the feature.
+
+- Define any new terminology.
+- Define the target persona: buildpack author, buildpack user, platform operator, platform implementor, and/or project contributor.
+- Explaining the feature largely in terms of examples.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing users and new users.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+A folder named `/platform/unsafe` should be reserved by specification.  Within that folder, an arbitrary collection of directories could be mounted (e.g. `/platform/unsafe/{gradle,maven}`).  Each of the mounts at those specified children would be ready/write instead of read only as all other mounts are configured.  Buildpacks have complete access during both `detect` and `build` and no other enforcment is required.
+
```

Comment:
The motivation makes sense. I'm trying to wrap my mind around how this directory/mount would be shared across multiple builds. I've written down a few thought experiments. Could you validate whether I'm thinking about this right?

### pack (local mount share)

In the case of something like `pack` it could make sense to **provide the same `unsafe` mount for every build**. This would enable the use case of: As a java developer with app A and app B, I don't want to have to download the same JDK or dependencies across builds.

Are there any concerns about sharing the same `unsafe` mount across multiple _very_ unrelated builds like a power dev that builds X amount of languages, uses various buildpacks/builders vendors? Or even more probable, using `pack` in a centralized CI system host like Jenkins.

### multi-tenant

For other multi-tenant platforms, they could theoretically share the same `unsafe` mount via additional configuration (such as specifying what additional directories inside `unsafe` to share) or a higher level mechanism (such as per "app", per "team", etc).

Reply:
@jromero

### `pack`

This wouldn't apply to dependencies that are included in buildpacks like the JDK addressed via #81.  Those have very different sharing and caching strategies (among other things, they're much safer to cache and share, and perform better when sharing is available) from what is being addressed by this PR, so you should focus on the downloading of application-specific dependencies.  The key use-case is reusing a common `.m2/repository` with tons of dependencies already downloaded in it.  This would be across unrelated single-language builds but typically not across languages.  It would also enable the use of locally-built snapshots of dependencies without needing to publish them to another location available to the build container.

Other than the same performance concerns of local mounting on MacOS generally, there are no concerns about mounting something broader than just a download cache.  You _could_ mount `$HOME` in there and while it wouldn't be intuitively useful (for example a `paketo-buildpacks/maven` would only be interested in seeing `/platform/unsafe/maven`), a custom buildpack built by a user could work to its heart's content.

### Multi-tenant

In the olden days (👴🏻) Jenkins used to run multiple builds, often simultaneously, on the same VM.  In these cases there *was* a shared `~/.m2/repository` and it **SUCKED**.  Concurrent builds against file systems where the build system assumed exclusive access is a recipe for failure.  To this day, it's unlikely that newer languages (like say Go) would hand this, and the necessary lag of going to an NFS, very well.  This was one of the foundational drivers behind the switch to building in containers across the entire industry.  So I don't see this is a feature for co-mingling multi-tenant builds as we'd re-introduce many of the same issues that containerization eliminated.

I'd argue that the `cache = true` layer for downloaded application dependencies in a CI system **is** sufficient in a way that it's not for local-development with something like `pack`.  In that case, the application is built slowly once, but then happens in an asynchronous way (triggered by a commit while you're off getting coffee), and you don't have the slowness of the first and the quickness of subsequent ones staring you in the face.  In addition, there is much more continuity between CI'd builds (i.e. you always build the same images over and over) leading to cache hits, rather than local development where you give images new, unique names all the time and the `cache = true` misses most of the time because it was for another "app" (image).

Reply:
This helps a lot to clarify, thank you!